### ISSUE TITLE
Drop redundant hashing step in FRI Merke tree building

### ISSUE THREAD

**Issue by aszepieniec** [TritonVM/triton-vm#176]
After discussing with @AlexanderLemmens it seems as though there is a redundant hashing step in the construction of FRI Merkle trees. The leaf is obtained from hashing an XFieldElement; the Merkle tree is then build out of such leafs. However, we can drop this step at no cost to security and instead start building the Merkle tree directly on top of these XFieldElements.

The next question is why we don't hash together 3 rather than 2 XFieldElements in the bottom layer. While this does generate a speedup for the prover, it also complicates matters a bit. We would need support for Merkle trees that are ternary at the bottom.

**Comment by sshine** [comment]
If I understand correctly, the XFieldElements that are hashed come from the codeword, which is the whole FRI domain subgroup. If we don't need to initially hash the `XFieldElement`s to form digests, can't we partition the FRI domain into `Digest`-sized chunks rather than `XFieldElement`-sized chunks and use the regular Merkle hashing on those? (Since `DIGEST_LENGTH` is not a power of 2, some padding would be necessary towards the end.)

**Comment by aszepieniec** [comment]
> the codeword, which is the whole FRI domain subgroup

The XFieldElements do come from the codeword, but the codeword is not the FRI domain. The codeword is the evaluation of some low degree polynomial on all elements of the FRI domain. (This FRI domain corresponds to a coset of a subgroup of order $2^k$.) Since this polynomials has coefficients living in the X-field, its evaluations also live in the X-field. So in the end the codeword consists of $3 \cdot 2^k$ B-field elements. You can partition it any way you like but at some point the verifier needs to read an X-field element with a given index, and it would be inconvenient if this one X-field element is spread out over two leafs, particularly if those two leafs are not siblings.

**Comment by sshine** [comment]
> it would be inconvenient if this one X-field element is spread out over two leafs, particularly if those two leafs are not siblings.

Would it be that inconvenient if those two leafs were siblings/neighbors?

```
                    ...                                         ...
                  /     \                                     /     \
[xa0 xa1 xa2 xb0 xb1] [xb2 xc0 xc1 xc2 xd0] [xd1 xd2 xe0 xe1 xe2] [xf0 xf1 xf2 xg0 xg1] xg2 ...
```

where `xa = [xa0 xa1 xa2]` is the first X-field element of the codeword, and so on.

If you hash the leafs by reading in the B-field elements backwards, I think you could find the root efficiently by hashing pairs of digests using stack memory entirely, without resorting to RAM.

**Comment by sshine** [comment]
> at some point the verifier needs to read an X-field element with a given index

Ah, I overlooked this part.

Must the X-field element with a given index correspond to a root in the tree for verifying that it's a member in the tree? In that case, I suppose that does make for extra book-keeping since an authentication path potentially needs to cover two leafs.

**Comment by aszepieniec** [comment]
> Must the X-field element with a given index correspond to a root in the tree for verifying that it's a member in the tree? In that case, I suppose that does make for extra book-keeping since an authentication path potentially needs to cover two leafs.

Exactly correct.

**Comment by jan-ferdinand** [comment]
Fixed in 2a9a9892.
